lib/kafka_ex/protocol/create_topics.ex

```diff
@@ -0,0 +1,145 @@
+  defmodule ReplicaAssignment do
+    defstruct partition: nil, replicas: nil
+    @type t :: %ReplicaAssignment{ partition: integer, replicas: [integer] }
+  end
+
+  defmodule ConfigEntry do
+    defstruct config_name: nil, config_value: nil
+    @type t :: %ConfigEntry{ config_name: binary, config_value: binary }
+  end
+
+  defmodule TopicRequest do
+    defstruct topic: nil,
+              num_partitions: -1,
+              replication_factor: -1,
+              replica_assignment: [],
+              config_entries: []
+    @type t :: %TopicRequest{
+      topic: binary,
+      num_partitions: integer,
+      replication_factor: integer,
+      replica_assignment: [ReplicaAssignment],
+      config_entries: [ConfigEntry],
+    }
+  end
+
+  defmodule Request do
+    @moduledoc false
+    defstruct create_topic_requests: nil, timeout: nil
+    @type t :: %Request{create_topic_requests: [TopicRequest], timeout: integer}
+  end
+
+  defmodule TopicError do
+    defstruct topic_name: nil, error_code: nil
+    @type t :: %TopicError{ topic_name: binary, error_code: integer }
+  end
+
+  defmodule Response do
+    @moduledoc false
+    defstruct topic_errors: nil
+    @type t :: %Response{topic_errors: [TopicError]}
+  end
+
+  @spec create_request(integer, binary, Request.t) :: binary
+  def create_request(correlation_id, client_id, create_topics_request) do
+    Protocol.create_request(:create_topics, correlation_id, client_id) <>
+      encode_topic_requests(create_topics_request.create_topic_requests) <>
+        << create_topics_request.timeout  :: 32-signed >>
+  end
+
+  @spec encode_topic_requests([TopicRequest.t]) :: binary
+  defp encode_topic_requests(requests) do
+    requests
+    |> map_encode(&encode_topic_request/1)
+  end
+
+  @spec encode_topic_request(TopicRequest.t) :: binary
+  defp encode_topic_request(request) do
+    encode_string(request.topic) <>
+      << request.num_partitions :: 32-signed, request.replication_factor :: 16-signed >> <>
+      encode_replica_assignments(request.replica_assignment) <>
+      encode_config_entries(request.config_entries)
+  end
+
+
+  @spec encode_replica_assignments([ReplicaAssignment.t]) :: binary
+  defp encode_replica_assignments(replica_assignments) do
+    replica_assignments |> map_encode(&encode_replica_assignment/1)
+  end
+
+  defp encode_replica_assignment(replica_assignment) do
+    << replica_assignment.partition :: 32-signed >> <>
+    replica_assignment.replicas |> map_encode(&(<< &1 :: 32-signed >>))
+  end
+
+  @spec encode_config_entries([ConfigEntry.t]) :: binary
+  defp encode_config_entries(config_entries) do
+    config_entries |> map_encode(&encode_config_entry/1)
+  end
+
+  @spec encode_config_entry(ConfigEntry.t) :: binary
+  defp encode_config_entry(config_entry) do
+    encode_string(config_entry.config_name) <> encode_nullable_string(config_entry.config_value)
+  end
+
+  @spec encode_nullable_string(String.t) :: binary
+  defp encode_nullable_string(text) do
```

Comment:
> nullable string could be a nil, but the spec doesn't allow that.

So I should leave it as it is ?

> This function could use pattern matching instead of an `if` as well.

I'll replace the `if` with pattern matching.

Reply:
by 'spec' I meant the typespec of the function: encode_nullable_string should take `nil | String.t`